Adds DeploymentConfig log functionality to Odo

cdrage opened the pull request:
This adds a new command `odo log` which will display the
DeploymentConfig logs for that particular component.

For example:

```sh
github.com/redhat-developer/odo  add-logs ✔                                                                                                                                                                                                                                                                                                                         15h55m
▶ ./odo list
ACTIVE     NAME        TYPE
*          nodejs3     nodejs

github.com/redhat-developer/odo  add-logs ✔                                                                                                                                                                                                                                                                                                                         15h57m
▶ ./odo log
Environment:
        DEV_MODE=false
        NODE_ENV=production
        DEBUG_PORT=5858
Launching via npm...
npm info it worked if it ends with ok
npm info using npm@3.10.9
npm info using node@v6.11.3
npm info lifecycle nodejs-ex@0.0.1~prestart: nodejs-ex@0.0.1
npm info lifecycle nodejs-ex@0.0.1~start: nodejs-ex@0.0.1

> nodejs-ex@0.0.1 start /opt/app-root/src
> node server.js

Server running on http://0.0.0.0:8080
```

cdrage commented:
Quite simple (only retrieving DeploymentConfig logs).

But I believe it's a step in the right direction.

What do you think @kadel ?

```sh
github.com/redhat-developer/odo  add-logs ✔                                                                                                                                                                                                                                                                                                                         15h55m  
▶ ./odo list
ACTIVE     NAME        TYPE
*          nodejs3     nodejs

github.com/redhat-developer/odo  add-logs ✔                                                                                                                                                                                                                                                                                                                         15h57m  
▶ ./odo log
Environment: 
        DEV_MODE=false
        NODE_ENV=production
        DEBUG_PORT=5858
Launching via npm...
npm info it worked if it ends with ok
npm info using npm@3.10.9
npm info using node@v6.11.3
npm info lifecycle nodejs-ex@0.0.1~prestart: nodejs-ex@0.0.1
npm info lifecycle nodejs-ex@0.0.1~start: nodejs-ex@0.0.1

> nodejs-ex@0.0.1 start /opt/app-root/src
> node server.js

Server running on http://0.0.0.0:8080
```

cdrage commented on pkg/occlient/occlient.go:
```diff
@@ -791,6 +795,33 @@ func (c *Client) FollowBuildLog(buildName string) error {
+	_, err = io.Copy(buf, rd)
+
+	// Print
+	fmt.Println(buf.String())
```
Okay, no worries. Should we change how to do https://github.com/redhat-developer/odo/blob/master/pkg/occlient/occlient.go#L1048 as well?

kadel replied:
We should, but let's do that in another PR.

cdrage commented:
Most changes corrected. Should we change https://github.com/redhat-developer/odo/blob/master/pkg/occlient/occlient.go#L1048 as well @kadel with regards to io.Writer? I'll implement it by passing in io.Writer from higher-up in the code.

kadel commented on cmd/log.go:
```diff
@@ -0,0 +1,58 @@
+
+		// Retrieve and set the current component if none passed
+		var currentComponent string
+		if len(args) == 0 {
```
yes, but if `inputComponent` is empty than it returns the current component, otherwise it checks if inputComponent exists.

This is exactly what you are doing here with `args[0]` no?

wouldn't this have the same result?
```go
currentComponent := ""
if len(args) == 1 {
  currentComponent = args[1]
}
currentComponent = getComponent(client, currentComponent, currentApplication, currentProject)

 

```

cdrage replied:
@kadel We're passing in blank for currentComponent.. which feels as if it's a undocumented function since it will return the active component rather than "getting" the component that's being passed in (currentComponent). Does that make sense?

kadel replied:
It is documented in a comment for that function.

I agree that it feels weird, but we have this logic in a lot of places, and it actually makes sense to have this abstracted away in one common function. Maybe a different name or small refactoring of that function might help.

kadel commented on pkg/occlient/occlient.go:
```diff
@@ -1020,7 +1021,7 @@ func (c *Client) WaitAndGetPod(selector string) (*corev1.Pod, error) {
 
 // FollowBuildLog stream build log to stdout
-func (c *Client) FollowBuildLog(buildName string) error {
+func (c *Client) FollowBuildLog(name string) error {
```
why was this changed from `buildName` to just `name`?

kadel commented on pkg/occlient/occlient.go:
```diff
@@ -1046,8 +1047,35 @@ func (c *Client) FollowBuildLog(buildName string) error {
+}
+
+// Display DeploymentConfig log to stdout
+func (c *Client) DisplayDeploymentConfigLog(buildName string) error {
```
the argument is called `buildName`  but this function accepts DeploymentConfig name as an argument

kadel commented:
> Most changes corrected. Should we change https://github.com/redhat-developer/odo/blob/master/pkg/occlient/occlient.go#L1048 as well @kadel with regards to io.Writer? I'll implement it by passing in io.Writer from higher-up in the code.

there already is an issue for that ;-) https://github.com/redhat-developer/odo/issues/312

cdrage commented:
@kadel Updated! I pass in an `io.Writer` now. I retrieve this using `color.Output` in order for us to get coloured output in Windows / macOS / Linux.

I have updated the function as well to write to io.Writer using `io.Copy`. This is all in yellow as well in order to differentiate and indicate to the user that this is raw OpenShift logs.

cdrage commented:
Let me know what you think regarding `getComponent` and I'll change it to your suggestion.

cdrage commented on pkg/component/component.go:
```diff
@@ -446,3 +446,17 @@ func GetComponentDesc(client *occlient.Client, currentComponent string, currentA
 }
+
+// Get Component logs
+func GetLogs(client *occlient.Client, applicationName string) (string, error) {
```
So for io.Writer, should we have this functionality higher-up, in the `cmd` package instead? At the moment, we have the "color" functionlity (using `color.Output`) within component.go rather than up in `cmd`. So perhaps I should move both.

kadel replied:
Not sure if I understand that :-(

There are functions that will print output in two colors.
For example `component.Watch`.
It is called from `cmd` package and it takes `io.Writer` as argument.
It writes to `io.Writer` messages in white and also in yellow. So the color can't be set in `cmd` package.

cdrage commented:
@kadel

I've done the following:

 - Added `os.Stdout` as a variable within `cmd/log.go` so we grab `io.Writer` at the most-highest abstraction level
 - Added `io.Writer` as parameters to both `GetLogs` as well as `DisplayDeploymentConfigLog`

kadel commented on cmd/log.go:
```diff
@@ -0,0 +1,62 @@
+
+		} else {
+
+			currentComponent = args[0]
```
we don't need this whole `else` section. This is already taken care of in `getComponent` function

cdrage commented:
I've also gone ahead and added a small integration test.

kadel commented on cmd/log.go:
```diff
@@ -0,0 +1,62 @@
+			// Get the current compoment
+			currentComponent = getComponent(client, "", currentApplication, currentProject)
+
+		} else {
```
this whole else section is not needed. This logic is already in `getComponent`.

The whole point of using `getComponent` is that you can replace this entire `if else` with  following 5 lines 

```go
currentComponent := ""
if len(args) == 1 {
  currentComponent = args[0]
}
currentComponent = getComponent(client, currentComponent, currentApplication, currentProject)

``

kadel commented on pkg/component/component.go:
```diff
@@ -446,3 +446,17 @@ func GetComponentDesc(client *occlient.Client, currentComponent string, currentA
+// Get Component logs
+func GetLogs(client *occlient.Client, applicationName string, stdout io.Writer) (string, error) {
+
+	// Retrieve the stdout pipe to pass to occlient
```
is this comment still valid?

cdrage commented:
@kadel Updated the comment :+1:

kadel commented on tests/e2e/e2e_test.go:
```diff
@@ -226,6 +226,12 @@ var _ = Describe("odoe2e", func() {
+
+			It("should be able to retrieve logs", func() {
+				runCmd("odo log")
+				runCmd("odo nodejs")
```
`odo nodejs` what is this command?

kadel commented on tests/e2e/e2e_test.go:
```diff
@@ -226,6 +226,12 @@ var _ = Describe("odoe2e", func() {
 			})
+
+			It("should be able to retrieve logs", func() {
+				runCmd("odo log")
```
this should probably test if something was returned

kadel commented on cmd/log.go:
```diff
@@ -0,0 +1,46 @@
+		currentProject := project.GetCurrent(client)
+
+		// Retrieve and set the currentComponent
+		currentComponent := getComponent(client, args[0], currentApplication, currentProject)
```
It is failing here if no argument is passed to `odo log` command.

If there is no argument, you have to pass an empty string as `inputComponent` (second) parameter and `getComponent` will return the name of the current component. If `inputComponent`  parameter is non-empty string it will check that `inputComponent`  exists and return it back.

kadel commented on pkg/component/component.go:
```diff
@@ -446,3 +446,15 @@ func GetComponentDesc(client *occlient.Client, currentComponent string, currentA
 }
+
+// Get Component logs
+func GetLogs(client *occlient.Client, applicationName string, stdout io.Writer) (string, error) {
```
We don't need to return `string`.

kadel commented:
There is something wrong with output color, I see some lines in yellow some in white.

<img width="1001" alt="screen shot 2018-05-15 at 10 52 36" src="https://user-images.githubusercontent.com/57206/40046844-370d61d0-582e-11e8-87f4-d56fec0f812c.png">

kadel commented:
I found out what is happening with the color output.
Looks like wildfly is also using colored output (prints WARN messages in yellow) and it resets our settings.

cdrage commented:
Updated!

Changes:

 - Fixed tests (tests check for exit codes anyways)
 - I found a nil error bug if there are *no* components that exist at all on OpenShift. I added this here: https://github.com/redhat-developer/odo/pull/429/files#diff-dace9c973da6801d7504aca65cf4da57R45 and within occlient here: https://github.com/redhat-developer/odo/pull/429/files#diff-6025e9b620e9321c03c11bf29e5c4dbdR1066)
 - Check to see if len(args) == 1 when picking a component (https://github.com/redhat-developer/odo/pull/429/files#diff-dace9c973da6801d7504aca65cf4da57R37)

kadel commented on pkg/occlient/occlient.go:
```diff
@@ -1052,6 +1053,43 @@ func (c *Client) FollowBuildLog(buildName string) error {
+		SubResource("log").Stream()
+
+	if rd == nil {
+		return errors.New("Unable to retrieve DeploymentConfig from OpenShift, does your component exist?")
```
We already have nice user-facing message in `cmd`.

cdrage commented:
@kadel got rid of the Uppercase! :+1:
